src/magellan_vision/scripts/line_detector.py

```diff
@@ -1,211 +1,156 @@
-                    # cv2.circle(crop, (x1, y1),(5), (0,255,0), 3)
-                    # cv2.circle(crop, (x2, y2),(5), (0,255,0), 3)
-            circles = crop
-            lines = crop2
-
-    if linesP is None or len(linesP) < 40:
-        skipped_frames += 1
-
-    cv2.imshow('Crop', crop)
-    # linesAndCircles = cv2.addWeighted(lines, 0.5, circles, 0.5,0)
-    # final = cv2.addWeighted(processed, 0.8, linesAndCircles, 0.5,0)
-    # cv2.imshow('Original', frame)
-    # cv2.imshow('Combo', final)
-    # print(skipped_frames)
-
-    # FIXME: what is this crap, please FIX ME
-    # plt.plot(np.unique(x_array_right),
-    #          np.poly1d(np.polyfit(x_array_right, y_array_right, 1))
-    #          (np.unique(x_array_right)))
-
-    # right_line_slope,right_line_intercept = best_fit(np.asarray(x_array_right),np.asarray(y_array_right))
-    # left_line_slope,left_line_intercept = best_fit(np.asarray(x_array_left),np.asarray(y_array_left))
-
-    toc = time.time()
-    delta_t = toc - tic
-    print("FPS: {}".format(1/delta_t))
-
-    key = cv2.waitKey(25)
-    if key == 27:
-        break
-
-video.release()
-cv2.destroyAllWindows
+
+from cv_bridge import CvBridge, CvBridgeError
+from sensor_msgs.msg import Image
+
+
+class PubSubNode(object):
+    def __init__(self):
+        # TODO FIX THIS
+        try:
+            # Global Constant
+            self._VERBOSE = rospy.get_param("VERBOSE")
+        except KeyError as e:
+            rospy.logerr("LineDetector: error in looking up param. {}".format(e))
+            raise
+
+        self._lock = threading.RLock()
+
+        '''Initialize ros publisher, ros subscriber'''
+        self._lines_object = Lines()
+        # topic where we publish
+        self._image_pub = rospy.Publisher("/perception/color/image_processed", Image, queue_size=5)
+        # self._point_arr_pub = rospy.Publisher('/perception/detected_points', Int32XYArr, queue_size=5)
+        self._bridge = CvBridge()
+
+        # subscribed Topic
+        self._subscriber = rospy.Subscriber("/camera/color/image_raw", Image, self._callback, queue_size=1)
+        self._image = None
+
+        if self._VERBOSE:
+            rospy.logdebug("subscribed to /camera/color/image_raw")
+
+    def run_detects(self):
+        with self._lock:
+            # Line detection
+            if self._image is not None:
+                image_np = self._lines_object.detect(self._image)
+
+                # conversion back to Image
+                try:
+                    ros_msg = self._bridge.cv2_to_imgmsg(image_np)
+                except CvBridgeError:
+                    rospy.logdebug('Could not convert image')
+                    return
+
+                # Publish Processed Image amnd Points
+                self._image_pub.publish(ros_msg)
+                # self._point_arr_pub.publish(intXYMsg)
+
+    '''Callback function of subscribed topic.
+    Here images get converted and features detected and published'''
```

Comment:
comment needs updated.

Reply:
I'm just going to remove this comment

Reply:
@asaba96 these are fixed just not pushed so I'm resolving